Mathlib/Algebra/Module/Torsion.lean

```diff
@@ -57,11 +57,49 @@ import Mathlib.RingTheory.SimpleModule.Basic
 * The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
   (in `M`).
 
+## TODO
+
+* Move the advanced material to a new file `RingTheory.Torsion`.
+* Replace `NoZeroSMulDivisors` with `Module.IsTorsionFree`
+
 ## Tags
 
 Torsion, submodule, module, quotient
 -/
 
+/-! ### Torsion-free modules -/
+
+namespace Module
+variable {R M : Type*} [Semiring R]
+
+section AddCommMonoid
+variable [AddCommMonoid M] [Module R M]
+
+variable (R M) in
+/-- A `R`-module `M` is torsion-free if scalar multiplication by an element `r : R` is injective if
+multiplication (on `R`) by `r` is.
+
+For domains, this is equivalent to the usual condition of `r • m = 0 → r = 0 ∨ m = 0`.
+TODO: Prove it. -/
+class IsTorsionFree where
+  isSMulRegular ⦃r : R⦄ : IsRegular r → IsSMulRegular M r
+
+instance [IsAddTorsionFree M] : IsTorsionFree ℕ M where
```

Comment:
Should we take `[Module ℕ M]` as argument?

Reply:
I'm enclined to do this sort of "non-generalisation" as the need arises. I know this sort of generalisation might be useful in representation theory, in particular in CFT, but let's wait until it actually is an issue.